src/contracts/libraries/GPv2Encoding.sol

```diff
@@ -373,29 +436,75 @@ library GPv2Encoding {
     /// @return owner The address of the signer.
     /// @return remainingCalldata Input calldata that has not been used to
     /// decode the current order.
-    function recoverEoaOwner(
+    function recoverErc712Signer(
         bytes32 domainSeparator,
         bytes32 orderDigest,
         bytes calldata encodedSignature
     ) internal pure returns (address owner, bytes calldata remainingCalldata) {
-        require(
-            encodedSignature.length >= EOA_SIGNATURE_STRIDE,
-            "GPv2: invalid encoding"
-        );
-
         bytes32 r;
         bytes32 s;
         uint8 v;
-        // NOTE: Use assembly to efficiently decode signature data.
+        (r, s, v, remainingCalldata) = decodeEcdsaSignature(encodedSignature);
+
+        // NOTE: Solidity allocates, but does not free, memory when:
+        // - calling the ABI encoding methods
+        // - calling the `ecrecover` precompile.
+        // However, we can restore the free memory pointer to before we made
+        // allocations to effectively free the memory. This is safe as the
+        // memory used can be discarded, and the memory pointed to by the free
+        // memory pointer **does not have to point to zero-ed out memory**.
+        // <https://docs.soliditylang.org/en/v0.7.6/internals/layout_in_memory.html>
+        uint256 freeMemoryPointer;
         // solhint-disable-next-line no-inline-assembly
         assembly {
-            // r = uint256(encodedTrade[0:32])
-            r := calldataload(encodedSignature.offset)
-            // s = uint256(encodedTrade[32:64])
-            s := calldataload(add(encodedSignature.offset, 32))
-            // v = uint8(encodedTrade[64])
-            v := shr(248, calldataload(add(encodedSignature.offset, 64)))
+            freeMemoryPointer := mload(0x40)
+        }
```

Comment:
Maybe this should be moved  around the signature recovery code so it covers all branches? I imagine the EIP-1271 case might need this as well.

Reply:
I also don't like the repetition, but I prefer to have it used the closest possible to where it's needed. We could have something like this:
```
function getFreeMemoryPointer() internal returns (uint256 pointer) {
        // comment....
        assembly {
            pointer := mload(0x40)
        }
}
function setFreeMemoryPointer(uint256 pointer) internal {
        // comment....
        assembly {
            mstore(0x40, pointer)
        }
}
```
and use them without comments:
```
{
        //...
        uint256 pointer = getFreeMemoryPointer()
        abi.encode(...);
        ecrecover(...);
        setFreeMemoryPointer(pointer);
        //...
}
```

Reply:
Really like this suggestion